php/installer/Database.class.inc

```diff
@@ -45,6 +45,163 @@ class Database
     }
 
     /**
+     * On executing `SHOW GRANT;`, rows with a single column are returned.
+     * Each row's value is a string that lists the MySQL user's privileges for
+     * a given table.
+     *
+     * @param string $raw The string of each row returned by `SHOW GRANT;`
+     *
+     * @see https://dev.mysql.com/doc/refman/5.7/en/show-grants.html
+     *
+     * @return object The parsed privilege for a string returned by `SHOW GRANT;`
+     */
+    private function _parsePrivilege($raw)
+    {
+        $token_arr = explode(" ON ", $raw);
+        if (count($token_arr) != 2) {
+            //Avoid letting the user see the "privileges", output to error_log()
+            error_log(
+                "Database::_parsePrivilege():Received {$raw}, could not split in two"
+            );
+            throw new Exception("Could not parse privilege");
+        }
+        $database_table_end = strpos($token_arr[1], " ");
+        $database_table     = substr($token_arr[1], 0, $database_table_end);
+
+        $privilege_arr_raw = substr($token_arr[0], strlen("GRANT "));
+        $privilege_arr     = explode(", ", $privilege_arr_raw);
+
+        return (object)array(
+                        "database_table" => $database_table,
+                        "privilege_arr"  => $privilege_arr,
+                       );
+    }
+    private $_privileges = null;
+    /**
+     * Fetches and caches all the current MySQL user's privileges.
+     * Subsequent calls return a cached copy.
+     *
+     * @see    https://dev.mysql.com/doc/refman/5.7/en/show-grants.html
+     * @return array An associative array mapping "`<database>`.`<table>`" strings
+     *               to an array of privileges
+     */
+    public function getOrFetchAllPrivileges()
+    {
+        if (is_null($this->_privileges)) {
+            $stmt = $this->PDO->query("SHOW GRANTS");
+            $arr  = $stmt->fetchAll(PDO::FETCH_COLUMN);
+            $stmt->closeCursor();
+
+            $this->_privileges = array();
+            foreach ($arr as $raw) {
+                $obj = $this->_parsePrivilege($raw);
+                $this->_privileges[$obj->database_table] = $obj->privilege_arr;
+            }
+        }
+        return $this->_privileges;
+    }
+    /**
```

Comment:
Is there a PHPCS sniff that can enforce that? It seems like something that should be caught with static analysis, not human review..